lib/remote/endpoint.hpp

```diff
@@ -61,6 +70,9 @@ class Endpoint final : public ObjectImpl<Endpoint>
 	mutable RingBuffer m_MessagesReceived{60};
 	mutable RingBuffer m_BytesSent{60};
 	mutable RingBuffer m_BytesReceived{60};
+
+	mutable std::shared_mutex m_MessageCountersMutex;
+	std::unordered_map<ApiFunction*, Atomic<uint_fast64_t>> m_MessageCounters;
```

Comment:
Why `ApiFunction*`? Didn't Julian introduced https://github.com/Icinga/icinga2/pull/10360/commits/500ad70b8cc89ce5c856bad91021ee76a6b84848, so that we don't have to do this anymore?

Reply:
I got it working only with a raw pointer to ApiFunction being forward-declared. (`class ApiFunction;`) Also, ApiFunction instances never vanish (see #10388).

Reply:
> I got it working only with a raw pointer to ApiFunction being forward-declared. (`class ApiFunction;`)

Ah, you could use `intrusive_ptr<ApiFunction>` then.

> Also, ApiFunction instances never vanish (see #10388).

Again, it's not about whether or not they vanish but https://github.com/Icinga/icinga2/commit/500ad70b8cc89ce5c856bad91021ee76a6b84848 was introduced so that we can use smart pointers for such cases as well like everywhere in our code base.

Reply:
> > I got it working only with a raw pointer to ApiFunction being forward-declared. (`class ApiFunction;`)
> 
> Ah, you could use `intrusive_ptr<ApiFunction>` then.

I know and already tried.

> > Also, ApiFunction instances never vanish (see #10388).
> 
> Again, it's not about whether or not they vanish but [500ad70](https://github.com/Icinga/icinga2/commit/500ad70b8cc89ce5c856bad91021ee76a6b84848) was introduced so that we can use smart pointers for such cases as well like everywhere in our code base.

Again, it's not about raw or smart pointers, but about ApiFunction, even with an include, not being completely defined at that point. And I don't see any **good** reason to go down this specific rabbit hole.

Reply:
> > > I got it working only with a raw pointer to ApiFunction being forward-declared. (`class ApiFunction;`)
> > 
> > 
> > Ah, you could use `intrusive_ptr<ApiFunction>` then.
> 
> I know and already tried.

And? Why did you change it then?

> even with an include, not being completely defined at that point

I changed it to `std::unordered_map<intrusive_ptr<ApiFunction>, Atomic<uint_fast64_t>> m_MessageCounters;` and Icinga 2 builds fine, so I don't know what you're talking about.

Reply:
Please share your complete diff (to my PR) and whether you use unity builds.

Reply:
```diff
diff --git a/lib/remote/endpoint.hpp b/lib/remote/endpoint.hpp
index 7d2de3f87..59084a579 100644
--- a/lib/remote/endpoint.hpp
+++ b/lib/remote/endpoint.hpp
@@ -72,7 +72,7 @@ private:
        mutable RingBuffer m_BytesReceived{60};

        mutable std::shared_mutex m_MessageCountersMutex;
-       std::unordered_map<ApiFunction*, Atomic<uint_fast64_t>> m_MessageCounters;
+       std::unordered_map<intrusive_ptr<ApiFunction>, Atomic<uint_fast64_t>> m_MessageCounters;
 };

 }
```

With and without unity build Icinga 2 builds fine.